File: correct-crypto/go/password_hash.go
package correctcrypto

import (
	"crypto/rand"
	"crypto/subtle"

	"golang.org/x/crypto/argon2"
)

// HashPassword derives an Argon2id hash with a fresh random salt.
func HashPassword(password []byte) ([]byte, error) {
	salt := make([]byte, 16)
	if _, err := rand.Read(salt); err != nil {
		return nil, err
	}
	key := argon2.IDKey(password, salt, 3, 64*1024, 4, 32) // time=3, mem=64MiB, threads=4
	return append(salt, key...), nil
}

// VerifyPassword recomputes the hash and compares it in constant time.
func VerifyPassword(stored, password []byte) bool {
	if len(stored) < 16 {
		return false
	}
	salt, expected := stored[:16], stored[16:]
	key := argon2.IDKey(password, salt, 3, 64*1024, 4, 32)
	return subtle.ConstantTimeCompare(key, expected) == 1
}

File: correct-crypto/rust/password_hash.rs
// Correct password hashing with Argon2id (argon2 crate).
use argon2::password_hash::rand_core::OsRng;
use argon2::password_hash::{PasswordHash, PasswordHasher, PasswordVerifier, SaltString};
use argon2::Argon2;

pub fn hash_password(password: &[u8]) -> String {
    let salt = SaltString::generate(&mut OsRng);        // random salt
    Argon2::default()
        .hash_password(password, &salt)
        .expect("hash failure")
        .to_string()
}

pub fn verify_password(encoded: &str, password: &[u8]) -> bool {
    match PasswordHash::new(encoded) {
        Ok(parsed) => Argon2::default().verify_password(password, &parsed).is_ok(),
        Err(_) => false,
    }
}
